x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java

```diff
@@ -0,0 +1,459 @@
+                                                try {
+                                                    return getHistogramValue(values.binaryValue());
+                                                } catch (IOException e) {
+                                                    throw new IllegalStateException("Cannot load doc value", e);
+                                                }
+                                            }
+                                        };
+                                    } catch (IOException e) {
+                                        throw new IllegalStateException("Cannot load doc values", e);
+                                    }
+
+                                }
+
+                                @Override
+                                public ScriptDocValues<?> getScriptValues() {
+                                    return new ScriptDocValues.Strings(getBytesValues());
+                                }
+
+                                @Override
+                                public SortedBinaryDocValues getBytesValues() {
+                                    try {
+                                        final BinaryDocValues values = DocValues.getBinary(context.reader(), fieldName);
+                                        return FieldData.singleton(values);
+                                    } catch (IOException e) {
+                                        throw new IllegalStateException("Cannot load doc values", e);
+                                    }
+                                }
+
+                                @Override
+                                public long ramBytesUsed() {
+                                    return 0; // Unknown
+                                }
+
+                                @Override
+                                public void close() {
+
+                                }
+                            };
+                        }
+
+                        @Override
+                        public AtomicHistogramFieldData loadDirect(LeafReaderContext context) throws Exception {
+                            return load(context);
+                        }
+
+                        @Override
+                        public SortField sortField(Object missingValue, MultiValueMode sortMode,
+                                                   XFieldComparatorSource.Nested nested, boolean reverse) {
+                            return null;
+                        }
+                    };
+                }
+
+                private HistogramValue getHistogramValue(final BytesRef bytesRef) throws IOException {
+                    final ByteBufferStreamInput streamInput = new ByteBufferStreamInput(
+                        ByteBuffer.wrap(bytesRef.bytes, bytesRef.offset, bytesRef.length));
+                    final int numValues = streamInput.readVInt();
```

Comment:
We could also avoid storing the length and consider the iterator exhausted when all bytes have been read.